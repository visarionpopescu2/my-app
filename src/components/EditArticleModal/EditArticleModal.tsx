import { Form, Input, Modal } from 'antd';
import { ARTICLE } from 'types';
import {useContext, useEffect, useState} from 'react';
import {editArticle} from 'actions';
import NotificationContext from 'context/NotificationContext/NotificationContext';

type ModalProps = {
  article: ARTICLE,
  isOpen: boolean,
  onClose: () => void,
  onSuccess: (article: ARTICLE) => void
}
function EditArticleModal(props: ModalProps) {
  const { article, isOpen, onClose, onSuccess } = props;
  const [isLoading, setIsLoading] = useState(false);
  const [form] = Form.useForm();
  const { openNotification } = useContext(NotificationContext);
  
  useEffect(() => {
    if (isOpen) {
      form.setFieldsValue({
        title: article.title,
        content: article.body
      });
    }
  }, [isOpen, form, article]);
  
  const handleClose = () => {
    form.resetFields();
    onClose();
  };
  
  const onSubmit = async () => {
    setIsLoading(true);
    let values;
    try {
      values = await form.validateFields();
    } catch (error) {
      console.log('Validate Failed:', error);
      setIsLoading(false);
      return;
    }
  
    const response = await editArticle({ ...article, body: values.content, title: values.title });
    if (response.error) {
      openNotification(response.error, 'error');
    } else {
      openNotification('The article has been successfully edited', 'success');
      onSuccess({ ...article, body: values.content, title: values.title });
      handleClose();
    }
    setIsLoading(false);
  };
  
  return (
    <Modal
      open={isOpen}
      title="Edit article"
      okText="Save"
      cancelText="Cancel"
      onCancel={handleClose}
      onOk={onSubmit}
      confirmLoading={isLoading}
    >
      <Form form={form} layout="vertical" requiredMark={false}>
        <Form.Item label="Title" name="title" rules={[{required: true, message: 'An article must have a title'}]}>
          <Input value={article.title} />
        </Form.Item>
        <Form.Item label="Content" name="content" rules={[{required: true, message: 'An article must have a body'}]}>
          <Input.TextArea value={article.body} />
        </Form.Item>
      </Form>
    </Modal>
  )
};

export default EditArticleModal;
