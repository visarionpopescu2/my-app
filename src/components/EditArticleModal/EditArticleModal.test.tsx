import { render } from '@testing-library/react';
import EditArticleModal from './EditArticleModal';

const ARTICLE = { userId: 1, id: 1, title: 'title', body: 'body' };
const onClose = jest.fn();
const onSuccess = jest.fn();

describe('EditArticleModal', () => {
  beforeEach(() => {
    Object.defineProperty(window, 'matchMedia', {
      writable: true,
      value: jest.fn().mockImplementation(query => ({
        matches: false,
        media: query,
        onchange: null,
        addListener: jest.fn(), // deprecated
        removeListener: jest.fn(), // deprecated
        addEventListener: jest.fn(),
        removeEventListener: jest.fn(),
        dispatchEvent: jest.fn(),
      })),
    });
  });

  it('should render EditArticleModal component when it is open', () => {
    let view = render(<EditArticleModal article={ARTICLE} isOpen={true} onClose={onClose} onSuccess={onSuccess}/>);
    expect(view.asFragment()).toMatchSnapshot();
  });
  
  it('should render EditArticleModal component when it is not open', () => {
    let view = render(<EditArticleModal article={ARTICLE} isOpen={false} onClose={onClose} onSuccess={onSuccess}/>);
    expect(view.asFragment()).toMatchSnapshot();
  });
});
