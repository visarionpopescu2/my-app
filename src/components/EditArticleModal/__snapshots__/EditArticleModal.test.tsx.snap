// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditArticleModal should render EditArticleModal component when it is not open 1`] = `<DocumentFragment />`;

exports[`EditArticleModal should render EditArticleModal component when it is open 1`] = `<DocumentFragment />`;
